crates/ra_assists/src/handlers/create_function.rs

```diff
@@ -0,0 +1,313 @@
+use ra_syntax::{
+    ast::{self, AstNode},
+    SmolStr, SyntaxKind, SyntaxNode, TextUnit,
+};
+
+use crate::{Assist, AssistCtx, AssistId};
+use ast::{CallExpr, Expr};
+use ra_fmt::leading_indent;
+
+// Assist: create_function
+//
+// Creates a stub function with a signature matching the function under the cursor.
+//
+// ```
+// fn foo() {
+//      bar<|>("", baz());
+// }
+//
+// ```
+// ->
+// ```
+// fn foo() {
+//      bar("", baz());
+// }
+//
+// fn bar(arg_1: &str, baz: Baz) {
+//      todo!();
+// }
+//
+// ```
+pub(crate) fn create_function(ctx: AssistCtx) -> Option<Assist> {
+    let path: ast::Path = ctx.find_node_at_offset()?;
+
+    if ctx.sema.resolve_path(&path).is_some() {
+        // The function call already resolves, no need to create a function
+        return None;
+    }
+
+    if path.qualifier().is_some() {
+        return None;
+    }
+
+    let call: ast::CallExpr = ctx.find_node_at_offset()?;
```

Comment:
We need to check that this call's `.receiver` is, in fact, `path`: `foo(b<|>ar)`.

Reply:
Do I understand correctly, in that case (`foo(b<|>ar)`), `path` will be `bar` but `CallExpr` will be the whole `foo(bar)` thing?

Reply:
I think so (but it's better to check)

Reply:
Gonna write a test :+1:

Reply:
that's what you were talking about, right?
https://github.com/rust-analyzer/rust-analyzer/pull/3746/files#diff-52ea394a636c2907a3e010b8d7e56a74R306
and you were right, this test currently fails.

Reply:
Fixed via https://github.com/rust-analyzer/rust-analyzer/pull/3746/files#diff-52ea394a636c2907a3e010b8d7e56a74R65
Is the whole `foo.syntax().descendents().find_map(ast::ChildI'mLookingFor::cast)` approach correct? It feels a bit dirty

Reply:
Yeah, I think it should be possible to do this without extra traversal. Here's one option I think:

```rust
    let path_expr: ast::PathExpr = ctx.find_node_at_offset();
    let call = expr.syntax().parent().and_then(ast::CallExpr::cast)?;
    let path = path_expr.path()?;
    if path.qualifier().is_some() {
        return None;
    }
```